heat/core/linalg/basics.py

```diff
@@ -1917,6 +1918,49 @@ def triu(m: DNDarray, k: int = 0) -> DNDarray:
 DNDarray.triu.__doc__ = triu.__doc__
 
 
+def vdot(x1: DNDarray, x2: DNDarray) -> DNDarray:
+    """
+    Computes the dot product of two vectors. Higher-dimensional arrays will be flattened.
+
+    Parameters
+    ----------
+    x1 : DNDarray
+        first input array. If it's complex, it's complex conjugate will be used.
+    x2 : DNDarray
+        second input array.
+
+    Raises
+    ------
+    ValueError
+        If the number of elements is inconsistent.
+
+    See Also
+    --------
+    dot
+            Return the dot product without using the complex conjugate.
+
+    Examples
+    --------
+    >>> a = ht.array([1+1j, 2+2j])
+    >>> b = ht.array([1+2j, 3+4j])
+    >>> ht.linalg.vdot(a,b)
+    DNDarray([(17+3j)], dtype=ht.complex64, device=cpu:0, split=None)
+    >>> ht.linalg.vdot(b,a)
+    DNDarray([(17-3j)], dtype=ht.complex64, device=cpu:0, split=None)
+    """
+    x1 = manipulations.flatten(x1)
+    x2 = manipulations.flatten(x2)
+
+    if x1.gnumel != x2.gnumel:
+        raise ValueError(
```

Comment:
`multiply` will check this again via `operations.__binary_op`, you can remove this.

Just for fun, I compared `if x1.gnumel != x2.gnumel:` to a `try: except:` check using `__torch_proxy__`, the latter is about 35% faster on my machine. Code:
```python
start = time.time()
for i in range(10000):
    if x1.gnumel != x2.gnumel:
        raise ValueError("error")        
end = time.time()
print("`if` check took ", end-start, "seconds.")

start = time.time()
for i in range(10000):
    try:
        torch.mul(x1.__torch_proxy__(), x2.__torch_proxy__())
    except RuntimeError: 
        raise ValueError("error")
end = time.time()
print("`try... except:` check took ", end-start, "seconds.")
```

Reply:
removed